crates/ruma-state-res/tests/it/resolve.rs

```diff
@@ -0,0 +1,341 @@
+}
+
+impl Event for Pdu {
+    type Id = OwnedEventId;
+
+    fn event_id(&self) -> &Self::Id {
+        &self.event_id
+    }
+
+    fn room_id(&self) -> &RoomId {
+        &self.room_id
+    }
+
+    fn sender(&self) -> &UserId {
+        &self.sender
+    }
+
+    fn origin_server_ts(&self) -> MilliSecondsSinceUnixEpoch {
+        self.origin_server_ts
+    }
+
+    fn event_type(&self) -> &TimelineEventType {
+        &self.kind
+    }
+
+    fn content(&self) -> &RawJsonValue {
+        &self.content
+    }
+
+    fn state_key(&self) -> Option<&str> {
+        self.state_key.as_deref()
+    }
+
+    fn prev_events(&self) -> Box<dyn DoubleEndedIterator<Item = &Self::Id> + '_> {
+        Box::new(self.prev_events.iter())
+    }
+
+    fn auth_events(&self) -> Box<dyn DoubleEndedIterator<Item = &Self::Id> + '_> {
+        Box::new(self.auth_events.iter())
+    }
+
+    fn redacts(&self) -> Option<&Self::Id> {
+        self.redacts.as_ref()
+    }
+}
+
+/// Extract `.content.room_version` from a PDU.
+#[derive(Deserialize)]
+struct ExtractRoomVersion {
+    room_version: RoomVersionId,
+}
+
+/// Type describing a resolved state event.
+#[derive(Serialize)]
+struct ResolvedState {
+    kind: StateEventType,
+    state_key: String,
+    event_id: OwnedEventId,
+
+    // Ignored in `PartialEq` and `Ord` because we don't want to consider it while sorting.
+    content: JsonValue,
+}
+
+impl PartialEq for ResolvedState {
+    fn eq(&self, other: &Self) -> bool {
+        self.kind == other.kind
+            && self.state_key == other.state_key
+            && self.event_id == other.event_id
+    }
+}
+
+impl Eq for ResolvedState {}
+
+impl Ord for ResolvedState {
+    fn cmp(&self, other: &Self) -> Ordering {
+        Ordering::Equal
+            .then(self.kind.cmp(&other.kind))
+            .then(self.state_key.cmp(&other.state_key))
+            .then(self.event_id.cmp(&other.event_id))
+    }
+}
+
+impl PartialOrd for ResolvedState {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// Information to be captured in snapshot assertions
+struct Snapshots {
+    /// The resolved state of the room.
+    resolved_state: BTreeSet<ResolvedState>,
+}
+
+/// Test a list of JSON files containing a list of PDUs and return the results.
+///
+/// State resolution is run both atomically for all PDUs and in batches of PDUs by file.
+fn run(paths: &[&str]) -> Snapshots {
+    // Run `cargo test -- --show-output` to view traces, set `RUST_LOG` to control filtering.
+    let _subscriber = tracing_subscriber::fmt()
+        .with_env_filter(EnvFilter::from_default_env())
+        .with_test_writer()
+        .finish()
+        .set_default();
+
+    // Wrapping a fallible function, as opposed to returning `Result` to the test harness, allows us
+    // to more easily control the output format of errors.
+    match try_run(paths) {
+        Ok(x) => x,
+        Err(e) => panic!("Error: {e}"),
```

Comment:
If we are going to panic anyway, why bother with errors? Let's just `.unwrap()`, `.expect()` and `panic!()` directly. It has the added benefit of showing where the error happened.

Reply:
`unwrap` and `expect` produce bad error messages and `?` is convenient. `panic!()` can be okay (as shown) but not returning `Result` encourages the use of `unwrap` and `expect`, which isn't great. Also, if we had any errors whose `source()` returned `Some`, this code could be modified to print out the source chain, which would be very useful.

Reply:
How are `unwrap` and `expect` worse than what we have now?

Reply:
`unwrap` provides no extra context and `expect` uses Debug instead of Display to print the error. This would, for example, make the atomic/batched diff unreadable because it would escape all the newlines.

Reply:
Several of the errors don't provide a context. I don't see how using "Debug" is considered "worse", it's just more verbose, which can be more helpful sometimes. I'm not sure why the atomic/batched diff is given as an example since it would not use `expect`.

Reply:
I've changed everything in `test_resolve` to use `expect` instead of returning a `Result`.